Fix #19855 - Allow non-SQL export plugins to export table structure

efturtle opened the pull request:
- Modified conditions on lines 822 and 826 in exportTable()
- Added OR clause to allow non-SQL plugins (PDF, Excel, etc.)
- Restores structure export for plugins without hasCreateTable()/hasCreateView()

The conditional logic in exportTable() was specific to ExportSql,
preventing other export plugins (PDF, Excel, etc.) from calling
exportStructure(). Modified conditions on lines 822 and 826 to allow
non-SQL plugins to export structure without requiring SQL-specific
methods like hasCreateTable() and hasCreateView().

Fixes #19855

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

efturtle committed:
Fix: Allow non-SQL export plugins to export table structure

- Modified conditions on lines 822 and 826 in exportTable()
- Added OR clause to allow non-SQL plugins (PDF, Excel, etc.)
- Restores structure export for plugins without hasCreateTable()/hasCreateView()

Signed-off-by: Alberto Cuevas <mexslacker@gmail.com>

liviuconcioiu commented:
Not sure about this change because

```PHP
if ($exportPlugin instanceof ExportSql && $exportPlugin->hasCreateView() || ! $exportPlugin instanceof ExportSql) {
```

is the same as

```PHP
if ($exportPlugin->hasCreateView()) {
```

efturtle commented:
Not necessarily.
The first part of the condition

```php
if ($exportPlugin instanceof ExportSql && $exportPlugin->hasCreateView())
``` 

checks whether the plugin is using ExportSql and has the “create view” flag enabled.

The new part added in this PR —
```php
|| ! $exportPlugin instanceof ExportSql
``` 
— ensures the condition also passes when the plugin is something other than ExportSql (like ExportPdf, ExportOdt, etc.).

Because of operator precedence in PHP, the `&&` part will be evaluated before the `||`, so it should behave correctly when SQL is selected in the dropdown for export.

efturtle committed:
fix: Allow non-SQL export plugins to export table structure
- Simplified conditions in line 821 & 826

Signed-off-by: Alberto Cuevas <mexslacker@gmail.com>

efturtle commented on src/Export/Export.php:
```diff
@@ -818,12 +818,12 @@ public function exportTable(
                     }
                 }
-            } elseif ($exportPlugin instanceof ExportSql && $exportPlugin->hasCreateTable()) {
+            } elseif ($exportPlugin instanceof ExportSql && $exportPlugin->hasCreateTable() || ! $exportPlugin instanceof ExportSql) {
```
Made new commit with simplified condition thank you @kamil-tekiela

williamdes reviewed (approved):
@kamil-tekiela are you okay with the new version ?

Old one: https://github.com/phpmyadmin/phpmyadmin/commit/038c59b1bf7cddd3634a665c864de976ba1eecf9

efturtle committed:
refactor: Use polymorphism for export plugin capabilities

Adds default hasCreateTable/View/Trigger/ProcedureFunction methods to
the ExportPlugin base class

Non-SQL plugins (PDF, etc.) now properly export table structure by
inheriting sensible defaults:
- hasCreateTable/hasCreateView return true by default
- hasCreateProcedureFunction/hasCreateTrigger return false (SQL only)

This fixes the bug where PDF exports couldn't show table structure
after the refactoring in 038c59b.

Signed-off-by: Alberto Cuevas <mexslacker@gmail.com>

williamdes reviewed (approved):
Thanks, this is much better to understand
Maybe the methods should be abstract
